hts.c

```diff
@@ -2507,15 +2507,39 @@ long long hts_parse_decimal(const char *str, char **strend, int flags)
     return (sign == '+')? n : -n;
 }
 
-const char *hts_parse_reg(const char *s, int *beg, int *end)
+/*
+ * A variant of hts_parse_reg which is reference-id aware.  It uses
+ * the iterator name2id callbacks to validate the region tokenisation works.
+ *
+ * This is necessary due to GRCh38 HLA additions which have reference names
+ * like "HLA-DRB1*12:17".
+ *
+ * On success the end of the reference is returned (colon or end of string).
+ * On failure NULL is returned, and if tid/getid are supplied *tid will be -1.
+ */
+const char *hts_parse_reg2(const char *s, int *tid, int *beg, int *end,
+                           hts_name2id_f getid, void *hdr)
 {
-    char *hyphen;
+    // FIXME: do we need to permit tid=-1 for reference "*" to indicate unmapped
+    // reads, and strictly have NULL as failure test?
+    int tid_, s_len = strlen(s); // int is sufficient given beg/end types
+    if (!tid) tid = &tid_;       // simplifies code below
```

Comment:
We should check all the pointers at the beginning, especially as the function would now be part of the interface.